File: ru/dencoder.py
raw_table = """а a
б '6
в v
г r
д g
е e
ё 'e
ж 'x
з z
и u
й 'u
к k
л l
м m
н h
о o
п n
р p
с c
т t
у y
ф 'q
х x
ц 'v
ч '4
ш w
щ 'w
ъ 'b
ы i
ь b
э '3
ю 'o
я q
' ''"""

table_ru_en = {y[0]: y[1] for y in [x.split(" ") for x in raw_table.split("\n")]}

table_en_ru = {v: k for k, v in table_ru_en.items()}

#print out encoding map
print("encode with:")
for k, v in table_ru_en.items():
	print(k, v)
print()

#print out decoding map
print("decode with:")
for k, v in sorted(table_en_ru.items()):
	print(k, v)
print()

def encode_ru(str):
	ret = []

	for char in str:
		if char in table_ru_en:
			ret.append(table_ru_en[char])
		else:
			ret.append(char)

	return "".join(ret)

def decode_ru(str):
	ret = []

	#for loop
	i = 0
	while i < len(str):
		if str[i] == "'":
			#invalid quote escape sequence
			if not i + 1 < len(str) or str[i:i+2] not in table_en_ru:
				return None
			ret.append(table_en_ru[str[i:i+2]])
			i += 1
		elif str[i] in table_en_ru:
			ret.append(table_en_ru[str[i]])
		else:
			ret.append(str[i])
		i += 1

	return "".join(ret)

File: ru/dencoder.js
let raw_table = `а a
б '6
в v
г r
д g
е e
ё 'e
ж 'x
з z
и u
й 'u
к k
л l
м m
н h
о o
п n
р p
с c
т t
у y
ф 'q
х x
ц 'v
ч '4
ш w
щ 'w
ъ 'b
ы i
ь b
э '3
ю 'o
я q
' ''`;

let table_ru_en = new Map();
let table_en_ru = new Map();

for (let it of raw_table.split("\n")){
	let parts = it.split(" ");
	table_ru_en.set(parts[0], parts[1]);
	table_en_ru.set(parts[1], parts[0]);
}

let encode_ru = (str) => {
	let ret = [];

	for (let char of str){
		if (table_ru_en.has(char)){
			ret.push(table_ru_en.get(char));
		}
		else{
			ret.push(char);
		}
	}
	return ret.join("");
};

let decode_ru = (str) => {
	let ret = [];

	for (let i = 0; i < str.length; ++i){
		if (str.charAt(i) == "'"){
			if (i + 1 >= str.length || !table_en_ru.has(str.substr(i, 2))){
				return null;
			}
			ret.push(table_en_ru.get(str.substr(i, 2)));
			++i;
		}
		else if (table_en_ru.has(str.charAt(i))){
			ret.push(table_en_ru.get(str.charAt(i)));
		}
		else{
			ret.push(str.charAt(i));
		}
	}
	return ret.join("");
};
